### PR TITLE
odoo: fix tests and update to 18.0

### PR DESCRIPTION
**Pull request by phanirithvij**
redo #344173

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by siriobalmelli — CHANGES_REQUESTED** [review]
I get a merge conflict with `master`:

```
Auto-merging pkgs/by-name/od/odoo15/package.nix
CONFLICT (content): Merge conflict in pkgs/by-name/od/odoo15/package.nix
Auto-merging pkgs/development/python-modules/aiohttp/default.nix
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by phanirithvij** [comment]
fixed the conflict, try again

**Comment by Erethon** [comment]
I tried this yesterday (before the newest changes) by cherry-picking the commits and applying them on master. It pretty much worked, Odoo 17 was functional, although some smaller features of it weren't working (like QR code generation), but I believe these weren't working in the past either. From my point of view, this fixes the issues of the existing package, so thank you!

**Comment by siriobalmelli** [comment]
Screenshot 2024-10-09 at 4 40 38 PM

... still running nixpkgs-review ... 🔢

**Comment by siriobalmelli** [comment]
Result of `nixpkgs-review pr 346397` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>stig</li>
    <li>stig.dist</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
I don't think the PR needs to target staging.

File: pkgs/development/python-modules/aiohttp/default.nix [hunk 1]
```diff
@@ -113,6 +114,7 @@ buildPythonPackage rec {
       # don't run benchmarks
       "test_import_time"
     ]
+    ++ lib.optionals isPy310 [ "test_https_proxy_unsupported_tls_in_tls" ]
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/aiohttp/default.nix:117]: Would be nice to have a short comment here

File: pkgs/by-name/od/odoo16/package.nix [hunk 2]
```diff
@@ -42,12 +22,9 @@ in python.pkgs.buildPythonApplication rec {
   src = fetchzip {
     url = "https://nightly.odoo.com/${odoo_version}/nightly/src/odoo_${version}.zip";
     name = "${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo16/package.nix:24]:
```suggestion
    name = "odoo-${version}";
```

File: pkgs/by-name/od/odoo16/package.nix [hunk 3]
```diff
@@ -4,34 +4,14 @@
 , rtlcss
 , wkhtmltopdf
 , nixosTests
+, odoo_version ? "16.0"
+, odoo_release ? "20240924"
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo16/package.nix:8]:
```suggestion
```
I don't think we should have those in inputs. The hashes are not updated so things don't work as expected and we generally have overrides for that.

File: pkgs/by-name/od/odoo15/package.nix [hunk 4]
```diff
@@ -48,12 +40,9 @@ in python.pkgs.buildPythonApplication rec {
   src = fetchzip {
     url = "https://nightly.odoo.com/${odoo_version}/nightly/src/odoo_${version}.zip";
     name = "${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo15/package.nix:42]:
```suggestion
    name = "odoo15-${version}";
```

File: pkgs/by-name/od/odoo15/package.nix [hunk 5]
```diff
@@ -17,27 +24,12 @@ let
           hash = "sha256-WnRbsy/PJcotZqY9mJPLadrYqkXykOVifLIbDyNf4s4=";
         };
 
+        nativeBuildInputs = [ self.setuptools ];
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo15/package.nix:27]:
```suggestion
        dependencies = [ self.setuptools ];
```

File: pkgs/by-name/od/odoo15/package.nix [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
-, nixosTests }:
+{ lib
+, fetchFromGitHub
+, fetchzip
+, python310
+, rtlcss
+, wkhtmltopdf
+, nixosTests
+, odoo_version ? "15.0"
+, odoo_release ? "20240924"
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo15/package.nix:9]:
```suggestion
```
same as below

File: pkgs/by-name/od/odoo/package.nix [hunk 7]
```diff
@@ -35,12 +35,9 @@ in python.pkgs.buildPythonApplication rec {
     # find latest version on https://nightly.odoo.com/${odoo_version}/nightly/src
     url = "https://nightly.odoo.com/${odoo_version}/nightly/src/odoo_${version}.zip";
     name = "${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo/package.nix:37]:
```suggestion
    name = "odoo-${version}";
```

File: pkgs/by-name/od/odoo/package.nix [hunk 8]
```diff
@@ -6,7 +6,7 @@
 , wkhtmltopdf
 , nixosTests
 , odoo_version ? "17.0"
-, odoo_release ? "20240610"
+, odoo_release ? "20240925"
```
Comment (SuperSandro2000) [inline pkgs/by-name/od/odoo/package.nix:9]: and same as below

**Comment by siriobalmelli** [comment]
> I don't think the PR needs to target staging.

That might explain how long it's taking me to build everything:

Screenshot 2024-10-10 at 12 12 08 AM

**Comment by SuperSandro2000** [comment]
Your rebuild is taking so long because you are targeting staging which has almost no cache. There is no reason to target staging because the rebuild count is almost nothing.

**Comment by siriobalmelli** [comment]
Result of `nixpkgs-review pr 346397` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>odoo</li>
    <li>odoo.dist</li>
    <li>odoo15</li>
    <li>odoo15.dist</li>
    <li>odoo16</li>
    <li>odoo16.dist</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
    <li>stig</li>
    <li>stig.dist</li>
  </ul>
</details>

**Comment by phanirithvij** [comment]
Seems like odoo has dropped 15.0 for their docker images and released 18 as well, see https://github.com/odoo/docker and https://github.com/docker-library/docs/tree/master/odoo

- I updated the target branch to master
- updated odoo to 18
- added `odoo17/package.nix`
- modified the update script to handle updating all versions not just the latest version
- made changes suggested by sandro

Should we remove odoo 15 or still keep it?

**Comment by mkg20001** [comment]
>Should we remove odoo 15 or still keep it?

Given that odoo is now at version 18 I'd say we should drop it on master. I think that's a breaking change according to releease scheudle? In that case we can simply drop it after release branch-off.

**Comment by mkg20001** [comment]
I'm seeing a lot of unresolved conversations for old diffs, please mark the resolved conversations as resolved. Also note that applying a code suggestion via github ui automatically marks the conversation as resolved and adds a commit to the branch (that can be squashed into others if it's not wanted as stand alone commit).

**Comment by siriobalmelli** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 346397`

---
### `aarch64-linux`
<details>
  <summary>:x: 10 packages failed to build:</summary>
  <ul>
    <li>odoo15</li>
    <li>odoo15.dist</li>
    <li>odoo16</li>
    <li>odoo16.dist</li>
    <li>odoo17</li>
    <li>odoo17.dist</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
    <li>stig</li>
    <li>stig.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>odoo</li>
    <li>odoo.dist</li>
  </ul>
</details>

**Comment by phanirithvij** [comment]
I tried `nix-build -A pkgsCross.aarch64-multiplatform.odoo15` on my x86_64-linux but can't do it with 16GB ram + 16GB swap.

File: pkgs/development/python-modules/aiohttp/default.nix [hunk 9]
```diff
@@ -113,6 +114,11 @@ buildPythonPackage rec {
       # don't run benchmarks
       "test_import_time"
     ]
+    # these tests fail with python310 but succeeds with 11+
+    ++ lib.optionals isPy310 [
```
Comment (mweinelt) [inline pkgs/development/python-modules/aiohttp/default.nix:118]: What about 3.9?

Comment (phanirithvij) [reply pkgs/development/python-modules/aiohttp/default.nix:118]:
```nix
$ for i in python3{9,10,11,12,13}Packages.aiohttp; do nix-build --no-out-link -A $i; done
       (stack trace truncated; use '--show-trace' to show the full trace)

       error: ipython-8.27.0 not supported for interpreter python3.9
/nix/store/j55np66p2kgj7m76g0kfbafb1wycxnnm-python3.10-aiohttp-3.10.5
/nix/store/i99cdds80c0f84bzsrmrwy2k6ik9crji-python3.11-aiohttp-3.10.5
/nix/store/i1csas85j560pbv182h9wkxlxqjscgqi-python3.12-aiohttp-3.10.5
       (stack trace truncated; use '--show-trace' to show the full trace)

       error: mypy-1.11.2 not supported for interpreter python3.13
```
py39 is already incompatible it seems?

**Comment by phanirithvij** [comment]
This is ready, but the update scripts are duplicated, since the pkgs/by-name doesn't allow accessing files outside the tree. https://github.com/NixOS/nixpkgs/pull/292214
I doubt we want to move these back to all-packages. So I think we have to live with this duplication.

Can everyone test it once, try to use odoo18 (if possible older versions too) for a few days and report any breakages here. I hope to get this in before 24.11 freeze. (< 2 weeks)

**Comment by siriobalmelli** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 346397`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>circus</li>
    <li>circus.dist</li>
    <li>odoo</li>
    <li>odoo.dist</li>
    <li>odoo15</li>
    <li>odoo15.dist</li>
    <li>odoo16</li>
    <li>odoo16.dist</li>
    <li>odoo17</li>
    <li>odoo17.dist</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
    <li>stig</li>
    <li>stig.dist</li>
  </ul>
</details>

**Comment by mkg20001** [comment]
Any objections to merge? Would merge in a few days

**Review by siriobalmelli — APPROVED** [review]

**Comment by siriobalmelli** [comment]
LGTM; I'll circle back to the aarch64-linux builds failures later as needed.

**Comment by phanirithvij** [comment]
Cross compilation works tho

```console
$ echo pkgsCross.aarch64-multiplatform.odoo{15,16,17,} | xargs -n1 nix build -f . --print-out-paths
/nix/store/7m2j0svg34vjnhqm8gyb21l3fz9fyyjm-odoo-15.0.20241010-aarch64-unknown-linux-gnu
/nix/store/68jbpcb3620kg5cw338zwbvax5yjzqnh-odoo-16.0.20241010-aarch64-unknown-linux-gnu
/nix/store/x0wq3s1dfbmwhgn519jw3mhxgizcmyh7-odoo-17.0.20241010-aarch64-unknown-linux-gnu
/nix/store/gagv0a46hbw8lg639y9fx0djjl67cksh-odoo-18.0.20241010-aarch64-unknown-linux-gnu
```

**Pull request merged**